File: src/main/java/org/hwabeag/hwaskyblock/events/UseEvent.java
package org.hwabeag.hwaskyblock.events;

import dev.lone.itemsadder.api.CustomBlock;
import dev.lone.itemsadder.api.ItemsAdder;
import org.bukkit.ChatColor;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Boat;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Minecart;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityMountEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.vehicle.VehicleDamageEvent;
import org.bukkit.event.vehicle.VehicleEntityCollisionEvent;
import org.bukkit.inventory.ItemStack;
import org.hwabeag.hwaskyblock.config.ConfigManager;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public class UseEvent implements Listener {

    FileConfiguration Config = ConfigManager.getConfig("setting");
    FileConfiguration SkyBlockConfig = ConfigManager.getConfig("skyblock");
    String Prefix = ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("hwaskyblock-system.prefix")));

    @EventHandler
    public void onPlayerDamageMinecart(VehicleDamageEvent event) {
        if (!(event.getVehicle() instanceof Minecart entity)) return;
        if (!(event.getAttacker() instanceof Player damager)) return;
        String name = damager.getName();
        World world = entity.getWorld();
        String world_name = world.getWorldFolder().getName();
        String[] number = world_name.split("\\.");
        if (Objects.equals(number[0], "HwaSkyBlock")) {
            String id = number[1];
            if (SkyBlockConfig.getString(id + ".leader") != null) {
                if (!Objects.equals(SkyBlockConfig.getString(id + ".leader"), name)) {
                    if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                        if (!SkyBlockConfig.getBoolean(id + ".break")) {
                            damager.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                            event.setCancelled(true);
                        }
                    } else {
                        if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.break")) {
                            damager.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                            event.setCancelled(true);
                        }
                    }
                }
            }
        }
    }

    @EventHandler
    public void onPlayerDamageBoat(VehicleDamageEvent event) {
        if (!(event.getVehicle() instanceof Boat entity)) return;
        if (!(event.getAttacker() instanceof Player damager)) return;
        String name = damager.getName();
        World world = entity.getWorld();
        String world_name = world.getWorldFolder().getName();
        String[] number = world_name.split("\\.");
        if (Objects.equals(number[0], "HwaSkyBlock")) {
            String id = number[1];
            if (SkyBlockConfig.getString(id + ".leader") != null) {
                if (!Objects.equals(SkyBlockConfig.getString(id + ".leader"), name)) {
                    if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                        if (!SkyBlockConfig.getBoolean(id + ".use.break")) {
                            damager.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                            event.setCancelled(true);
                        }
                    } else {
                        if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.break")) {
                            damager.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                            event.setCancelled(true);
                        }
                    }
                }
            }
        }
    }

    @EventHandler
    public void onEntityCollision(VehicleEntityCollisionEvent event) {
        Entity vehicle = event.getVehicle();
        Entity entity = event.getEntity();
        if (entity instanceof Player player && vehicle instanceof Minecart) {
            String name = player.getName();
            World world = entity.getWorld();
            String world_name = world.getWorldFolder().getName();
            String[] number = world_name.split("\\.");
            if (Objects.equals(number[0], "HwaSkyBlock")) {
                String id = number[1];
                if (SkyBlockConfig.getString(id + ".leader") != null) {
                    if (!Objects.equals(SkyBlockConfig.getString(id + ".leader"), name)) {
                        if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                            if (!SkyBlockConfig.getBoolean(id + ".use.minecart")) {
                                player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                event.setCancelled(true);
                                return;
                            }
                        } else {
                            if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.minecart")) {
                                player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                event.setCancelled(true);
                                return;
                            }
                        }
                    }
                }
            }
        }
        if (entity instanceof Player player && vehicle instanceof Boat) {
            String name = player.getName();
            World world = entity.getWorld();
            String world_name = world.getWorldFolder().getName();
            String[] number = world_name.split("\\.");
            if (Objects.equals(number[0], "HwaSkyBlock")) {
                String id = number[1];
                if (SkyBlockConfig.getString(id + ".leader") != null) {
                    if (!Objects.equals(SkyBlockConfig.getString(id + ".leader"), name)) {
                        if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                            if (!SkyBlockConfig.getBoolean(id + ".use.boat")) {
                                player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                event.setCancelled(true);
                            }
                        } else {
                            if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.boat")) {
                                player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                event.setCancelled(true);
                            }
                        }
                    }
                }
            }
        }
    }

    @EventHandler
    public void onEntityMount(EntityMountEvent event) {
        Entity entity = event.getEntity();
        Entity vehicle = event.getMount();
        if (entity instanceof Player player && vehicle instanceof Minecart) {
            String name = player.getName();
            World world = entity.getWorld();
            String world_name = world.getWorldFolder().getName();
            String[] number = world_name.split("\\.");
            if (Objects.equals(number[0], "HwaSkyBlock")) {
                String id = number[1];
                if (SkyBlockConfig.getString(id + ".leader") != null) {
                    if (!Objects.equals(SkyBlockConfig.getString(id + ".leader"), name)) {
                        if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                            if (!SkyBlockConfig.getBoolean(id + ".use.minecart")) {
                                player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                event.setCancelled(true);
                                return;
                            }
                        } else {
                            if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.minecart")) {
                                player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                event.setCancelled(true);
                                return;
                            }
                        }
                    }
                }
            }
        }
        if (entity instanceof Player player && vehicle instanceof Boat) {
            String name = player.getName();
            World world = entity.getWorld();
            String world_name = world.getWorldFolder().getName();
            String[] number = world_name.split("\\.");
            if (Objects.equals(number[0], "HwaSkyBlock")) {
                String id = number[1];
                if (SkyBlockConfig.getString(id + ".leader") != null) {
                    if (!Objects.equals(SkyBlockConfig.getString(id + ".leader"), name)) {
                        if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                            if (!SkyBlockConfig.getBoolean(id + ".use.boat")) {
                                player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                event.setCancelled(true);
                            }
                        } else {
                            if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.boat")) {
                                player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                event.setCancelled(true);
                            }
                        }
                    }
                }
            }
        }
    }

    @EventHandler
    public void onInteract(PlayerInteractEvent event) {
        Player player = event.getPlayer();
        String name = player.getName();
        if (event.getAction() == Action.RIGHT_CLICK_BLOCK) {
            Block block = event.getClickedBlock();
            World world = Objects.requireNonNull(block).getWorld();
            String world_name = world.getWorldFolder().getName();
            String[] number = world_name.split("\\.");
            if (Objects.equals(number[0], "HwaSkyBlock")) {
                String id = number[1];
                if (SkyBlockConfig.getString(id + ".leader") != null) {
                    if (!Objects.equals(SkyBlockConfig.getString(id + ".leader"), name)) {
                        CustomBlock customBlock = CustomBlock.byAlreadyPlaced(block);
                        if (customBlock != null) {
                            ItemStack CustomCrop = ItemsAdder.getCustomBlock(block);
                            @NotNull String CustomCrop_DisplayName = CustomCrop.getItemMeta().getDisplayName();
                            for (String type : Objects.requireNonNull(Config.getConfigurationSection("Custom-Crop-Interact")).getKeys(false)) {
                                if (CustomCrop_DisplayName.contains(type)) {
                                    if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                        if (!SkyBlockConfig.getBoolean(id + ".use.farm")) {
                                            player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                            event.setCancelled(true);
                                            return;
                                        }
                                    } else {
                                        if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.farm")) {
                                            player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                            event.setCancelled(true);
                                            return;
                                        }
                                    }
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("CHEST")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.chest")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.chest")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("TRAPDOOR")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.trapdoor")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.trapdoor")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("DOOR")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.door")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.door")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("BARREL")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.barrel")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.barrel")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("HOPPER")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.hopper")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.hopper")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("FURNACE")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.furnace")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.furnace")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("BLAST")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.furnace")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.furnace")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("SHULKER")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.shulker")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.shulker")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                            }
                            return;
                        }
                        if (block.getType().name().contains("BUTTON")) {
                            if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                if (!SkyBlockConfig.getBoolean(id + ".use.button")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                                if (!SkyBlockConfig.getBoolean(id + ".use.door")) {
                                    for (int x = -1; x <= 1; x++) {
                                        for (int y = -1; y <= 1; y++) {
                                            for (int z = -1; z <= 1; z++) {
                                                if (x == 0 && y == 0 && z == 0) continue;
                                                Block adjacentBlock = block.getRelative(x, y, z);
                                                if (adjacentBlock.getType().name().contains("DOOR")) {
                                                    event.setCancelled(true);
                                                    return;
                                                }
                                            }
                                        }
                                    }
                                }
                                if (!SkyBlockConfig.getBoolean(id + ".use.trapdoor")) {
                                    for (int x = -1; x <= 1; x++) {
                                        for (int y = -1; y <= 1; y++) {
                                            for (int z = -1; z <= 1; z++) {
                                                if (x == 0 && y == 0 && z == 0) continue;
                                                Block adjacentBlock = block.getRelative(x, y, z);
                                                if (adjacentBlock.getType().name().contains("TRAPDOOR")) {
                                                    event.setCancelled(true);
                                                    return;
                                                }
                                            }
                                        }
                                    }
                                }
                            } else {
                                if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.button")) {
                                    player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                    event.setCancelled(true);
                                    return;
                                }
                                if (!SkyBlockConfig.getBoolean(id + ".use.door")) {
                                    for (int x = -1; x <= 1; x++) {
                                        for (int y = -1; y <= 1; y++) {
                                            for (int z = -1; z <= 1; z++) {
                                                if (x == 0 && y == 0 && z == 0) continue;
                                                Block adjacentBlock = block.getRelative(x, y, z);
                                                if (adjacentBlock.getType().name().contains("DOOR")) {
                                                    event.setCancelled(true);
                                                    return;
                                                }
                                            }
                                        }
                                    }
                                }
                                if (!SkyBlockConfig.getBoolean(id + ".use.trapdoor")) {
                                    for (int x = -1; x <= 1; x++) {
                                        for (int y = -1; y <= 1; y++) {
                                            for (int z = -1; z <= 1; z++) {
                                                if (x == 0 && y == 0 && z == 0) continue;
                                                Block adjacentBlock = block.getRelative(x, y, z);
                                                if (adjacentBlock.getType().name().contains("TRAPDOOR")) {
                                                    event.setCancelled(true);
                                                    return;
                                                }
                                            }
                                        }
                                    }
                                }
                                return;
                            }
                            if (block.getType().name().contains("ANVIL")) {
                                if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                    if (!SkyBlockConfig.getBoolean(id + ".use.anvil")) {
                                        player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                        event.setCancelled(true);
                                        return;
                                    }
                                } else {
                                    if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.anvil")) {
                                        player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                        event.setCancelled(true);
                                        return;
                                    }
                                }
                                return;
                            }
                            if (block.getType().name().contains("BEACON")) {
                                if (SkyBlockConfig.getString(id + ".sharer." + name) == null) {
                                    if (!SkyBlockConfig.getBoolean(id + ".use.beacon")) {
                                        player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                        event.setCancelled(true);
                                    }
                                } else {
                                    if (!SkyBlockConfig.getBoolean(id + ".sharer." + name + ".use.beacon")) {
                                        player.sendActionBar(Prefix + ChatColor.translateAlternateColorCodes('&', Objects.requireNonNull(Config.getString("message-event.no_permission"))));
                                        event.setCancelled(true);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    @EventHandler
    public void onDamage(EntityDamageByEntityEvent event) {
        if (!(event.getEntity() instanceof Player entity)) return;
        World world = entity.getWorld();
        String world_name = world.getWorldFolder().getName();
        String[] number = world_name.split("\\.");
        if (Objects.equals(number[0], "HwaSkyBlock")) {
            String id = number[1];
            if (SkyBlockConfig.getString(id + ".leader") != null) {
                if (!SkyBlockConfig.getBoolean(id + ".pvp")) {
                    event.setCancelled(true);
                }
            }
        }
    }
}

File: src/main/kotlin/events/player/UseEvent.kt
package org.hwabaeg.hwaskyblock.events.player

import dev.lone.itemsadder.api.CustomBlock
import dev.lone.itemsadder.api.ItemsAdder
import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.ChatColor
import org.bukkit.block.Block
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.entity.Boat
import org.bukkit.entity.Entity
import org.bukkit.entity.Minecart
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.EntityMountEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.vehicle.VehicleDamageEvent
import org.bukkit.event.vehicle.VehicleEntityCollisionEvent
import org.hwabaeg.hwaskyblock.database.DatabaseManager
import org.hwabaeg.hwaskyblock.database.config.ConfigManager
import java.util.*

class UseEvent : Listener {
    var Config: FileConfiguration = ConfigManager.getConfig("setting")!!
    var MessageConfig: FileConfiguration = ConfigManager.getConfig("message")!!
    var Prefix: String = ChatColor.translateAlternateColorCodes(
        '&',
        Objects.requireNonNull<String?>(Config.getString("hwaskyblock-system.prefix"))
    )

    fun sendNoPermissionActionBar(player: Player) {
        val message = ChatColor.translateAlternateColorCodes(
            '&',
            Prefix + Objects.requireNonNull<String?>(MessageConfig.getString("message-event.no_permission"))
        )
        player.spigot().sendMessage(ChatMessageType.ACTION_BAR, TextComponent(message))
    }

    @EventHandler
    fun onPlayerDamageMinecart(event: VehicleDamageEvent) {
        val vehicle = event.vehicle
        if (vehicle !is Minecart) return
        if (event.attacker !is Player) return

        val damager = event.attacker as Player
        val name = damager.name
        val world = vehicle.world
        val worldName = world.worldFolder.name
        val number = worldName.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        if (number[0] == "HwaSkyBlock") {
            val id = number[1]
            val leader = DatabaseManager.getSkyBlockData(id, "getSkyBlockLeader") as? String
            if (leader != null && leader != name) {
                val permissionKey = "$id.sharer.$name"
                val isSharer = DatabaseManager.getSkyBlockData(id, "getSkyBlockShare") != null
                val useBreakPermission = if (!isSharer) {
                    DatabaseManager.getSkyBlockData(id, "getSkyBlockBreak") as? Boolean ?: false
                } else {
                    DatabaseManager.getSkyBlockData(id, "getSkyBlockShareBreak") as? Boolean
                        ?: false
                }

                if (useBreakPermission != true) {
                    sendNoPermissionActionBar(damager)
                    event.isCancelled = true
                }
            }
        }
    }

    @EventHandler
    fun onPlayerDamageBoat(event: VehicleDamageEvent) {
        val vehicle = event.vehicle
        if (vehicle !is Boat) return

        val damager = event.attacker
        if (damager !is Player) return

        val name = damager.name
        val world = vehicle.world
        val worldName = world.worldFolder.name
        val parts = worldName.split('.')

        if (parts.size < 2 || parts[0] != "HwaSkyBlock") return

        val id = parts[1]

        if (DatabaseManager.getSkyBlockData(id, "getSkyBlockLeader") == null) return

        val isSharer = DatabaseManager.getShareList(id).contains(name)

        val hasBreakPermission = if (isSharer)
            DatabaseManager.getShareData(id, name, "isUseBreak") as? Boolean ?: false
        else
            DatabaseManager.getSkyBlockData(id, "getSkyBlockBreak") as? Boolean ?: false

        if (!hasBreakPermission) {
            sendNoPermissionActionBar(damager)
            event.isCancelled = true
        }
    }


    @EventHandler
    fun onEntityCollision(event: VehicleEntityCollisionEvent) {
        val vehicle: Entity = event.vehicle
        val entity = event.entity

        if (entity is Player && (vehicle is Minecart || vehicle is Boat)) {
            val name = entity.name
            val world = entity.world
            val worldName = world.worldFolder.name
            val parts = worldName.split('.')
            if (parts[0] != "HwaSkyBlock" || parts.size < 2) return

            val id = parts[1]

            val leader = DatabaseManager.getSkyBlockData(id, "getSkyBlockLeader")
            if (leader == null || leader == name) return

            val isSharer = DatabaseManager.getShareList(id).contains(name)

            if (vehicle is Minecart) {
                val hasPermission = if (isSharer)
                    DatabaseManager.getShareData(id, name, "isUseMinecart") as? Boolean ?: false
                else
                    DatabaseManager.getSkyBlockData(id, "getSkyBlockMinecart") as? Boolean ?: false

                if (!hasPermission) {
                    sendNoPermissionActionBar(entity)
                    event.isCancelled = true
                }
            }
            if (vehicle is Boat) {
                val hasPermission = if (isSharer)
                    DatabaseManager.getShareData(id, name, "isUseBoat") as? Boolean ?: false
                else
                    DatabaseManager.getSkyBlockData(id, "getSkyBlockBoat") as? Boolean ?: false

                if (!hasPermission) {
                    sendNoPermissionActionBar(entity)
                    event.isCancelled = true
                }
            }
        }
    }


    @EventHandler
    fun onEntityMount(event: EntityMountEvent) {
        val entity = event.entity
        val vehicle = event.mount

        if (entity is Player && (vehicle is Minecart || vehicle is Boat)) {
            val name = entity.name
            val world = entity.world
            val worldName = world.worldFolder.name
            val parts = worldName.split('.')
            if (parts.size < 2 || parts[0] != "HwaSkyBlock") return

            val id = parts[1]
            val leader = DatabaseManager.getSkyBlockData(id, "getSkyBlockLeader")
            if (leader == null || leader == name) return

            val isSharer = DatabaseManager.getShareList(id).contains(name)

            val hasPermission = when (vehicle) {
                is Minecart -> if (isSharer)
                    DatabaseManager.getShareData(id, name, "isUseMinecart") as? Boolean ?: false
                else
                    DatabaseManager.getSkyBlockData(id, "getSkyBlockMinecart") as? Boolean ?: false

                is Boat -> if (isSharer)
                    DatabaseManager.getShareData(id, name, "isUseBoat") as? Boolean ?: false
                else
                    DatabaseManager.getSkyBlockData(id, "getSkyBlockBoat") as? Boolean ?: false

                else -> true
            }

            if (!hasPermission) {
                sendNoPermissionActionBar(entity)
                event.isCancelled = true
            }
        }
    }


    @EventHandler
    fun onInteract(event: PlayerInteractEvent) {
        val player = event.player
        val name = player.name
        if (event.action == Action.RIGHT_CLICK_BLOCK) {
            val block = event.clickedBlock ?: return
            val world = block.world
            val worldName = world.worldFolder.name
            val number = worldName.split('.')

            if (number[0] == "HwaSkyBlock") {
                val id = number[1]
                val leader =
                    DatabaseManager.getSkyBlockData(id.toString(), "getSkyBlockLeader") as? String
                if (leader != null && leader != name) {
                    val customBlock = CustomBlock.byAlreadyPlaced(block)
                    if (customBlock != null) {
                        handleCustomBlockInteraction(player, block, id, name, event)
                    } else {
                        handleStandardBlockInteraction(player, block, id, name, event)
                    }
                }
            }
        }
    }

    fun handleCustomBlockInteraction(
        player: Player,
        block: Block,
        id: String,
        name: String,
        event: PlayerInteractEvent
    ) {
        val customCrop = ItemsAdder.getCustomBlock(block)
        val customCropDisplayName = customCrop.itemMeta!!.displayName

        for (type in Config.getConfigurationSection("Custom-Crop-Interact")?.getKeys(false) ?: emptySet()) {
            if (customCropDisplayName.contains(type)) {
                val isSharer = DatabaseManager.getShareList(id).contains(name)
                val hasFarmPermission = if (isSharer)
                    DatabaseManager.getShareData(id, name, "isUseFarm") as? Boolean ?: false
                else
                    DatabaseManager.getSkyBlockData(id, "getSkyBlockFarm") as? Boolean ?: false

                if (!hasFarmPermission) {
                    sendNoPermissionActionBar(player)
                    event.isCancelled = true
                    return
                }
            }
        }
    }

    fun handleStandardBlockInteraction(
        player: Player,
        block: Block,
        id: String,
        name: String,
        event: PlayerInteractEvent
    ) {
        val blockType = block.type.name
        val blockPermissions = mapOf(
            "CHEST" to "chest",
            "TRAPDOOR" to "trapdoor",
            "DOOR" to "door",
            "BARREL" to "barrel",
            "HOPPER" to "hopper",
            "FURNACE" to "furnace",
            "BLAST" to "furnace",
            "SHULKER" to "shulker",
            "BUTTON" to "button",
            "ANVIL" to "anvil",
            "BEACON" to "beacon"
        )

        for ((blockName, permission) in blockPermissions) {
            if (blockType.contains(blockName)) {
                checkPermissionAndCancel(player, id, name, permission, event)
                return
            }
        }
    }

    fun checkPermissionAndCancel(
        player: Player,
        id: String,
        name: String,
        permission: String,
        event: PlayerInteractEvent
    ) {
        val isSharer = DatabaseManager.getShareList(id).contains(name)
        if (!isSharer) {
            if (!(DatabaseManager.getSkyBlockData(id, "getSkyBlockPermission") as? Boolean
                    ?: false)
            ) {
                sendNoPermissionActionBar(player)
                event.isCancelled = true
            }
        } else {
            if (!(DatabaseManager.getShareData(
                    id,
                    name,
                    null
                ) as? Boolean ?: false)
            ) {
                sendNoPermissionActionBar(player)
                event.isCancelled = true
            }
        }
    }

    @EventHandler
    fun onDamage(event: EntityDamageByEntityEvent) {
        val entity = event.entity
        if (entity !is Player) return

        val world = entity.world
        val worldName = world.worldFolder.name
        val parts = worldName.split(".")

        if (parts.size >= 2 && parts[0] == "HwaSkyBlock") {
            val id = parts[1]
            if (DatabaseManager.getSkyBlockData(id, "getSkyBlockLeader") != null) {
                if (!(DatabaseManager.getSkyBlockData(id, "getSkyBlockPvP") as? Boolean ?: false)) {
                    event.isCancelled = true
                }
            }
        }
    }

}